joblib/parallel.py

```diff
@@ -1699,11 +1713,7 @@ def _retrieve(self):
                 time.sleep(0.01)
                 continue
 
-            # We need to be careful: the job list can be filling up as
-            # we empty it and Python list are not thread-safe by
-            # default hence the use of the lock
-            with self._lock:
-                batched_results = self._jobs.popleft()
+            batched_results = self._jobs.popleft()
```

Comment:
I suspectt that this lock is not needed since deque pop and append methods _are_ thread-safe, but we didn't try to remove it yet. Yet it must be removed  in order to make the example work, since we need threads to be able to concurrently retrieve results and submit new tasks, to enable a feedback loop.